### ISSUE TITLE
Still too hard to install.

### ISSUE THREAD

**Issue by darron** [octohost/octohost#124]
I think I can solve this quite a bit easier if we actually use Docker better:

`Build the damn thing with Docker containers that build and launch other Docker containers.`

Not totally sure how best to do this - I think the containers I will need are:

1. Consul container - has the metadata and health checks for the containers.
2. nginx container w/consul template - routes and reverse proxies all the containers.
3. gitreceive container - gets the pushes and builds the containers.

There are some challenges:

1. Persistent templates and configuration files for Consul.
2. Persistent templates and configuration files for nginx.
3. Where to the git pushes go?

If we just mount the local filesystem for the git and configuration files, then that makes the containers not very stateful - but that feels wrong for some reason.

Thoughts?

**Comment by derek-adair** [comment]
This is absolutely wonderful.  +1,000,000.

**Comment by derek-adair** [comment]
So, this is a pretty wild idea.  But i believe those challenges can be overcome entirely by leveraging hashicorps [terraform](http://terraform.io/).  In my experience, leveraging git and git-recieve to deploy to octo was the biggest conceptual hurdle.  To me, it feels hack-ish.  While it definitely meets the minimum bar for what it does, I feel like the UX of using terraform, and the different benefits (such as a state file, representing your current infrastructure), maket his a no-brainer.

**But derek, how?**

A [**A custom octo provider**](https://terraform.io/docs/plugins/provider.html)!  Boy howdy, it would bring me joy to deploy my apps to my own private heroku service with `terraform apply`.  

Benefits: 

1. This would give us one command to not only [rebuild the consul config](https://github.com/octohost/octohost/issues/65), but rebuild entire docker containers on demand (via [terraform taint](https://terraform.io/docs/commands/taint.html)).

2. I am pretty sure this can be used to bring zero downtime](https://github.com/octohost/octohost/issues/65) to octo, although my measely brain hasn't quite groked that yet.

3. Clustering.  Oh-Em-Gee.  If i could specify a count in my terraform template and spin up multiple nodes that i can load balance..... /drool.

Basically i Feel like octo has outgrown git-recieve.  This will remove a large amount of complexity from this project, and essentially "outsource" the problems of manipulating infrastructure to something that does it... REALLY FUCKING WELL.

**Comment by derek-adair** [comment]
I'm very serious about this terraform provider regardless of how it pertains to this particular issue, i've been conceptualizing this for a while now and have discussed some superficial project details with @joshuacox.  Wasn't sure if it was worth filing an issue quite yet, but depending on your thoughts here I can do that.

**Comment by joshuacox** [comment]
Just chiming in here, I still very much like the gitreceive part, I always want to be able to merely push a repo at it and have an entire build happen right before my eyes.

That being said a handy dandy web interface that showed status, and maybe let you orchestrate containers a bit more.  For example marking a container to be built externally on another clustering mechanism.  I'm not certain something like flynn.io, controlcenter.io or deis.io might be the way to go here, I'm looking into lot's of options in this field right now (it looks like it most certainly will have a .io tld;).  

and dockerizing the individual tasks seems like a very natural next step, eventually you might even get everything in docker containers and move the host vm into coreOS or something similarly lightweight instead of ubuntu.

**Comment by darron** [comment]
Yeah. I like the git receive a ton too. But I would like to be able to just pull containers and launch them without having to push source code if possible. 
 
Having config data on all hosts through Consul is solved. 

If we adjust the Consul service registration to use something like progrium's registrator that could make multiple octohosts much easier. 

Getting a scheduler that can place containers smartly on multiple octo hosts is the most complex part I think. There's a whole bunch of options though. 

I want to both simplify the deploy procedure as well as make it a little more scalable with additional docker hosts. 

There's so much work going on with docker plugins right now. If we simplify and go closer to a stock Docker install, we may be able to use docker machine/swarm etc. 

Thanks for the feedback! I may start a v2 branch after the next release. Just waiting for Docker 1.6 and Consul 0.5.1 - should be soon.

**Comment by derek-adair** [comment]
@joshuacox i understand the allure of a simple mechanism like git-recieve.  Its a minimal solution with a bunch of flexibility.  I might say *too* flexible.  I see a potential pain point for users in having to resolve merges in order to push.  Its not an issue for those of us who are experienced in git, more specifically if you have ever managed a repository.  It raises the minimum bar for adoption of Octohost to those who can merge.  If thats something we decide is worth conceding thats fine, i suppose.  I'd just like to remind you that the major hurdle for me  with octo was figuring out WTF git-recieve was and how to use it.  Even after you alleviated the key generating pain point, conceptually i believe it obfuscates what octo does and prevented me from seeing the value of this system.

How i believe it should work, is just like any other of the "hubs" out there, where I can set up a git hook and trigger a build.  This build should be exposed through said fancy web UI, and tagged with a version of some kind.

Personally, i do not give a single fuck on how this is achieved.  If i have a central repository of "artifacts"[1], that insulates users from merge issues, I would be very very happy.  Maybe that means making terraform talking with git-recieve? Maybe we make "Octo-hub" and keep git.  Maybe i'm smoking way too much crack[2] and none of this makes sense!

1: Artifact-  pre-built images that will allow me to spin up trusted containers on-demand
2: I Dont actually smoke crack.

**Comment by joshuacox** [comment]
I think if we had a fancy web interface to pull containers/artifacts/vms/etc from an external registry it would bring a ton of usability to those people who don't understand why you'd use vim instead of M$ word, much less what plaintext is, and even further less what source code management is.  This is a major reason why lately I've been working on some docker containers to give a more tangible interface on status and other whizbang dashboards with kibana.  I want to let the non CLI junkies join in on the fun too.

That being said the gitreceive is not too open.  I've certainly gotten used to throwing stuff in consul keys for config type stuff.  But I still think there might be value in adding a config file into a repo, making the gitreceive even more powerful, possibly telling the gitreceive to fork on finding it and do something else entirely from the normal flow (of course we can set this possible imaginary feature in a consul var as well).  For example, a rudimentary multicontainer repo could be uploaded with multiple docker files specified in config up top and octo would fire off all the specified dockerfiles.  Or up the ante and have it call fig instead, or docker compose, or fire up a docker swarm, or send it to other octohosts to be built and mirrored there as well, or something else entirely.

**Comment by NigelThorne** [comment]
It's interesting that dokku moved away from git-recieve. They just use a git plugin now. (I believe)

**Comment by darron** [comment]
@NigelThorne I'd be really surprised if you couldn't push git repos to dokku. Even Deis still does that.

I am going to try and get something that can install a cluster - like Terraform - because of what I've been able to do with [Docker Swarm](https://github.com/octohost/www/issues/6#issuecomment-95068509)

Between using Atlas to build the image and using Terraform (or something like it) to deploy it - it just might be almost push button.

**Comment by NigelThorne** [comment]
You can git push to dokku. They just don't use git-receive any more. So they don't have a git account any more. etc.
